src/encoder.rs

```diff
@@ -2,12 +2,50 @@
 use std::cmp::min;
 use std::io;
 use std::io::prelude::*;
+use std::fmt;
+use std::error;
 
 use weezl::{BitOrder, encode::Encoder as LzwEncoder};
 
 use crate::traits::{WriteBytesExt};
 use crate::common::{Block, Frame, Extension, DisposalMethod};
 
+#[derive(Debug)]
+/// Encoding error.
+pub enum EncodingError {
+    /// The image has too many colors.
+    TooManyColors,
+    /// The image has no color palette which is required.
+    MissingColorPalette,
+    /// Wraps `std::io::Error`.
+    Io(io::Error),
+}
```

Comment:
Those are unlikely to remain the only errors in particular if extensions are to be supported more exhaustively. Adding more should not be a breaking change. I don't think we want to use `#[non_exhaustive]`, introduced in `1.36`, and thus bump MSRV to one incompatible with current `image`. Instead, I'd propose the design the be more opaque so that we can change the inner representation at least. Maybe see [`image`'s error for some inspiration](https://github.com/image-rs/image/blob/master/src/error.rs)?

Reply:
Ye this definitely seems like the better approach. On that note should I try and apply the same treatment to the `DecodingError` while I'm at it? I would believe having both error kinds be structured the same way is probably the better choice.

Reply:
Consistency between the two error types would be good, yes. I can see three basic approaches:

```rust
enum EncodingError {
    TooManyColors,
    MissingColorPalette,
    Io(io::Error),
    __NonExhaustive,
}
```

```rust
struct FormatError {
    // private
}

enum EncodingError {
    Format(FormatError)
    Io(io::Error)
}
```

or
```rust
struct EncodingError {
    // private
}

impl EncodingError {
    fn as_io(&self) -> Option<&io::Error>;
}
```

Reply:
I'd say the first one seems unreasonable when looking at the DecodingError, as it would have way too many variants I believe.
I would probably lean towards the second approach, will probably be able to make a clearer decision if I take a better look at the error cases again in a few days.